server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java

```diff
@@ -2069,6 +2249,235 @@ private List<TsFileResource> getFileResourceListForQuery(
     return tsfileResourcesForQuery;
   }
 
+  /**
+   * alter timeseries encoding & compressionType<br>
+   * 1、flush and close tsfile<br>
+   * 2、locks<br>
+   * 3、write temp tsfiles<br>
+   * 4、unregister old tsfiles and release locks<br>
+   * 5、rename temp tsfiles<br>
+   * 6、register tsfiles<br>
+   */
+  public void alter(
+      PartialPath fullPath, TSEncoding curEncoding, CompressionType curCompressionType)
+      throws IOException {
+
+    final String logKey =
+        logicalStorageGroupName + "-" + dataRegionId + "-" + fullPath.getFullPath();
+    // If there are still some old version tsfiles, the delete won't succeeded.
+    if (upgradeFileCount.get() != 0) {
+      throw new IOException(
+          "Alter failed. " + "Please do not delete until the old files upgraded.");
+    }
+    if (SettleService.getINSTANCE().getFilesToBeSettledCount().get() != 0) {
+      throw new IOException("Alter failed. " + "Please do not delete until the old files settled.");
+    }
+    logger.info("[alter timeseries] {} syncCloseAllWorkingTsFileProcessors", logKey);
+    // flush & close
+    syncCloseAllWorkingTsFileProcessors();
```

Comment:
Add dataRegion write lock before closing all working processors to avoid writing operation from users later.